### PR TITLE
WIP: feat(Observer.ts): add observer DP

### PR DESCRIPTION
**Pull request by RoeiRom**

### CODE REVIEW

File: client/src/components/App/Content/InvestigationForm/TabManagement/ClinicalDetails/ClinicalDetails.tsx [hunk 1]
```diff
@@ -45,6 +46,12 @@ const ClinicalDetails: React.FC = (): JSX.Element => {
         setHasBackgroundDiseases, setSymptoms, setBackgroundDiseases, context, setIsolationCityName, setIsolationStreetName, setStreetsInCity
     });
 
+    React.useEffect(() => {
+        tabsObserver.subscribe(id, () => {
```
Comment (NaamanTsur) [inline client/src/components/App/Content/InvestigationForm/TabManagement/ClinicalDetails/ClinicalDetails.tsx:50]: You need to cleanup the event subscription when the component leaves the DOM. Maybe call it `unsubscribe`.

File: client/src/commons/ObserverDP/Observer.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+    subscribe(id: number, gotObserverUpdate: () => void) {
+        this._observants.set(id, gotObserverUpdate);
+    }
+
+    notifyAll() {
+        this._observants.forEach(notification => notification());
+    }
+
+    notifySpecific(id: number) {
```
Comment (NaamanTsur) [inline client/src/commons/ObserverDP/Observer.ts:17]: I think calling the function `notify` is clear enough, considering there is also a `notifyAll` function.

File: client/src/components/App/Content/InvestigationForm/InvestigationForm.tsx [hunk 3]
```diff
@@ -81,7 +84,9 @@ const InvestigationForm: React.FC = (): JSX.Element => {
                             <div className={classes.buttonSection}>
                                 <PrimaryButton test-id={currentTab.id === LAST_TAB_ID ? 'endInvestigation' : 'continueToNextStage'}
                                     onClick={() => {
-                                        currentTab.id === LAST_TAB_ID ? confirmFinishInvestigation(epidemiologyNumber) : handleSwitchTab();
+                                        tabsObserver.notifySpecific(currentTab.id);
+                                        setCurrentTab(tabs[(currentTab.id + 1) % tabs.length]);
+                                        //currentTab.id === LAST_TAB_ID ? confirmFinishInvestigation(epidemiologyNumber) : handleSwitchTab();
```
Comment (NaamanTsur) [inline client/src/components/App/Content/InvestigationForm/InvestigationForm.tsx:89]: Remove commented code.

File: client/src/components/App/Content/InvestigationForm/TabManagement/ExposuresAndFlights/ExposuresAndFlights.tsx [hunk 4]
```diff
@@ -26,6 +27,12 @@ const ExposuresAndFlights = () => {
     const [departureDate, setDepartureDate] = React.useState<Date>();
     const [arrivalDate, setArrivalDate] = React.useState<Date>();
 
+    React.useEffect(() => {
+        tabsObserver.subscribe(id, () => {
```
Comment (NaamanTsur) [inline client/src/components/App/Content/InvestigationForm/TabManagement/ExposuresAndFlights/ExposuresAndFlights.tsx:31]: Same issue here with event cleanup.

File: client/src/components/App/Content/InvestigationForm/TabManagement/PersonalInfoTab/PersonalInfoTab.tsx [hunk 5]
```diff
@@ -66,6 +67,9 @@ const PersonalInfoTab: React.FC = (): JSX.Element => {
 
     React.useEffect(()=> {
         fetchPersonalInfo();
+        tabsObserver.subscribe(id, () => {
```
Comment (NaamanTsur) [inline client/src/components/App/Content/InvestigationForm/TabManagement/PersonalInfoTab/PersonalInfoTab.tsx:70]: Same issue here with event cleanup.

**Review by NaamanTsur — CHANGES_REQUESTED** [review]
Honestly I still don't have an understanding of the project to know whether the `observer` pattern is useful here so I won't comment on it at the moment.

File: client/src/components/App/Content/InvestigationForm/TabManagement/TabManagement.tsx [hunk 6]
```diff
@@ -23,13 +23,13 @@ export const tabs: TabObj[] = [
         id: 1,
         name: TabNames.CLINICAL_DETAILS,
         isDisabled: false,
-        displayComponent: <ClinicalDetails />
+        displayComponent: <ClinicalDetails id={1} />
```
Comment (sandymash) [inline client/src/components/App/Content/InvestigationForm/TabManagement/TabManagement.tsx:26]: could take id from existing TabObj id

**Pull request closed without merging**